File: NextcloudTalk/NCConnectionController.m
/**
 * SPDX-FileCopyrightText: 2020 Nextcloud GmbH and Nextcloud contributors
 * SPDX-License-Identifier: GPL-3.0-or-later
 */

#import "NCConnectionController.h"

#import "AFNetworking.h"

#import "NCAPIController.h"
#import "NCDatabaseManager.h"
#import "NCSettingsController.h"
#import "NCUserInterfaceController.h"

#import "NextcloudTalk-Swift.h"

NSString * const NCAppStateHasChangedNotification           = @"NCAppStateHasChangedNotification";
NSString * const NCConnectionStateHasChangedNotification    = @"NCConnectionStateHasChangedNotification";

@implementation NCConnectionController

@synthesize appState        = _appState;
@synthesize connectionState = _connectionState;

+ (NCConnectionController *)sharedInstance
{
    static dispatch_once_t once;
    static NCConnectionController *sharedInstance;
    dispatch_once(&once, ^{
        sharedInstance = [[self alloc] init];
    });
    return sharedInstance;
}

- (id)init
{
    self = [super init];
    if (self) {
        
        self.appState = kAppStateUnknown;
        self.connectionState = kConnectionStateUnknown;
        
        [self checkAppState];
        
        [[AFNetworkReachabilityManager sharedManager] setReachabilityStatusChangeBlock:^(AFNetworkReachabilityStatus status) {
            NSLog(@"Reachability: %@", AFStringFromNetworkReachabilityStatus(status));
            [self checkConnectionState];
        }];
    }
    return self;
}

- (void)notifyAppState
{
    NSDictionary *userInfo = [NSDictionary dictionaryWithObject:@(self.appState) forKey:@"appState"];
    [[NSNotificationCenter defaultCenter] postNotificationName:NCAppStateHasChangedNotification
                                                        object:self
                                                      userInfo:userInfo];
}

- (void)notifyConnectionState
{
    NSDictionary *userInfo = [NSDictionary dictionaryWithObject:@(self.connectionState) forKey:@"connectionState"];
    [[NSNotificationCenter defaultCenter] postNotificationName:NCConnectionStateHasChangedNotification
                                                        object:self
                                                      userInfo:userInfo];
}

- (BOOL)isNetworkAvailable {
    return [AFNetworkReachabilityManager sharedManager].reachable;
}

- (void)checkConnectionState
{
    if (![self isNetworkAvailable]) {
        [self setConnectionState:kConnectionStateDisconnected];
        [self notifyConnectionState];
    } else {
        ConnectionState previousState = self.connectionState;
        [self setConnectionState:kConnectionStateConnected];
        [self checkAppState];
        if (previousState == kConnectionStateDisconnected) {
            [self notifyConnectionState];
        }
    }
}

- (void)checkAppState
{
    TalkAccount *activeAccount = [[NCDatabaseManager sharedInstance] activeAccount];
    SignalingSettings *activeAccountSignalingConfig = [[[NCSettingsController sharedInstance] signalingConfigurations] objectForKey:activeAccount.accountId];

    if (!activeAccount.server || !activeAccount.user) {
        [self setAppState:kAppStateNotServerProvided];
        [[NCUserInterfaceController sharedInstance] presentLoginViewController];
    } else if (!activeAccount.userId || !activeAccount.userDisplayName) {
        [self setAppState:kAppStateMissingUserProfile];
        [[NCSettingsController sharedInstance] getUserProfileForAccountId:activeAccount.accountId withCompletionBlock:^(NSError *error) {
            if (error) {
                [self notifyAppState];
            } else {
                [self checkAppState];
            }
        }];
    } else if (!activeAccountSignalingConfig) {
        [self setAppState:kAppStateMissingServerCapabilities];
        [[NCSettingsController sharedInstance] getCapabilitiesForAccountId:activeAccount.accountId withCompletionBlock:^(NSError *error) {
            if (error) {
                [self notifyAppState];
                return;
            }

            [self setAppState:kAppStateMissingSignalingConfiguration];
            [[NCSettingsController sharedInstance] updateSignalingConfigurationForAccountId:activeAccount.accountId withCompletionBlock:^(NCExternalSignalingController * _Nullable signalingServer, NSError *error) {
                if (error) {
                    [self notifyAppState];
                    return;
                }

                [self checkAppState];
            }];
        }];
    } else {
        // Fetch additional data asynchronously.
        // We set the app as ready, so we don’t need to wait for this to complete.
        [[NCSettingsController sharedInstance] getUserGroupsAndTeamsForAccountId:activeAccount.accountId];
        [self setAppState:kAppStateReady];
    }
    
    [self notifyAppState];
}


@end

File: NextcloudTalk/NCConnectionController.swift
//
// SPDX-FileCopyrightText: 2020 Nextcloud GmbH and Nextcloud contributors
// SPDX-License-Identifier: GPL-3.0-or-later
//

import Foundation

extension Notification.Name {
    static let NCAppStateHasChangedNotification = Notification.Name(rawValue: "NCAppStateHasChangedNotification")
    static let NCConnectionStateHasChangedNotification = Notification.Name(rawValue: "NCConnectionStateHasChangedNotification")
}

@objc extension NSNotification {
    public static let NCAppStateHasChangedNotification = Notification.Name.NCAppStateHasChangedNotification
    public static let NCConnectionStateHasChangedNotification = Notification.Name.NCConnectionStateHasChangedNotification
}

@objc enum AppState: Int {
    case unknown = 0
    case noServerProvided
    case missingUserProfile
    case missingServerCapabilities
    case missingSignalingConfiguration
    case ready
}

@objc public enum ConnectionState: Int {
    case unknown = 0
    case disconnected
    case connected
}

@objcMembers
class NCConnectionController: NSObject {

    static let shared = NCConnectionController()

    public var appState: AppState = .unknown
    public var connectionState: ConnectionState = .unknown

    override init() {
        super.init()

        self.checkAppState()

        AFNetworkReachabilityManager.shared().setReachabilityStatusChange { status in
            print("Reachability: \(AFStringFromNetworkReachabilityStatus(status))")
            self.checkConnectionState()
        }
    }

    internal func notifyAppState() {
        // Use NSNumber for objc compatibility
        let dict = ["appState": NSNumber(value: self.appState.rawValue)]
        NotificationCenter.default.post(name: .NCAppStateHasChangedNotification, object: self, userInfo: dict)
    }

    internal func notifyConnectionState() {
        // Use NSNumber for objc compatibility
        let dict = ["connectionState": NSNumber(value: self.connectionState.rawValue)]
        NotificationCenter.default.post(name: .NCConnectionStateHasChangedNotification, object: self, userInfo: dict)
    }

    public func checkConnectionState() {
        if !AFNetworkReachabilityManager.shared().isReachable {
            self.connectionState = .disconnected
            self.notifyConnectionState()
        } else {
            let previousState = self.connectionState
            self.connectionState = .connected
            self.checkAppState()

            if previousState == .disconnected {
                self.notifyConnectionState()
            }
        }
    }

    public func checkAppState() {
        if NCDatabaseManager.sharedInstance().numberOfAccounts() == 0 {
            self.appState = .noServerProvided
            NCUserInterfaceController.sharedInstance().presentLoginViewController()
            self.notifyAppState()

            return
        }

        let activeAccount = NCDatabaseManager.sharedInstance().activeAccount()
        let signalingConfig = NCSettingsController.sharedInstance().signalingConfigurations.object(forKey: activeAccount.accountId)

        if activeAccount.user.isEmpty || activeAccount.userDisplayName.isEmpty {
            self.appState = .missingUserProfile

            NCSettingsController.sharedInstance().getUserProfile(forAccountId: activeAccount.accountId) { error in
                if error != nil {
                    self.notifyAppState()
                } else {
                    self.checkAppState()
                }
            }
        } else if signalingConfig == nil {
            self.appState = .missingServerCapabilities

            NCSettingsController.sharedInstance().getCapabilitiesForAccountId(activeAccount.accountId) { error in
                if error != nil {
                    self.notifyAppState()
                    return
                }

                self.appState = .missingSignalingConfiguration

                NCSettingsController.sharedInstance().updateSignalingConfiguration(forAccountId: activeAccount.accountId) { _, error in
                    if error != nil {
                        self.notifyAppState()
                        return
                    }

                    self.checkAppState()
                }
            }
        } else {
            // Fetch additional data asynchronously.
            // We set the app as ready, so we don’t need to wait for this to complete.
            NCSettingsController.sharedInstance().getUserGroupsAndTeams(forAccountId: activeAccount.accountId)
            self.appState = .ready
        }

        self.notifyAppState()
    }

}
